plugin/pkg/admission/jsonpath/admission.go

```diff
@@ -0,0 +1,211 @@
+// TODO: This is copied from pkg/kubectl/custom_column_printer.go move into the pkg/util/jsonpath package.
+func massageJSONPath(pathExpression string) (string, error) {
+	if len(pathExpression) == 0 {
+		return pathExpression, nil
+	}
+	submatches := jsonRegexp.FindStringSubmatch(pathExpression)
+	if submatches == nil {
+		return "", fmt.Errorf("unexpected path string, expected a 'name1.name2' or '.name1.name2' or '{name1.name2}' or '{.name1.name2}'")
+	}
+	if len(submatches) != 3 {
+		return "", fmt.Errorf("unexpected submatch list: %v", submatches)
+	}
+	var fieldSpec string
+	if len(submatches[1]) != 0 {
+		fieldSpec = submatches[1]
+	} else {
+		fieldSpec = submatches[2]
+	}
+	return fmt.Sprintf("{.%s}", fieldSpec), nil
+}
+
+type JSONPathAdmissionRule struct {
+	// Version indicates the version for the rule.  For now this has to be "v1"
+	Version string
+
+	// Name is the name of this rule
+	Name string
+
+	// KindRegexp holds a regular expression used to determine which Kinds of
+	// objects this rule applies to.
+	KindRegexp string `json:"kindRegexp", yaml:"kindRegexp"`
+
+	// Path holds a JSONPath expression to select particular fields
+	Path string `json:"path", yaml:"path"`
+
+	// MatchRegexp holds a matching expression, all values returned by the JSONPath expression (above)
+	// must match this expression for the object to be admitted
+	MatchRegexp string `json:"matchRegexp", yaml:"matchRegexp"`
+
+	// AcceptEmptyMatch indicates if an empty match is acceptable.  Empty match means that the
+	// JSONPath expression above returns no values.  Default is false, meaning that at least one
+	// value that matches MatchRegexp is required for admission
+	AcceptEmptyMatch bool `json:"acceptEmptyMatch", yaml:"acceptEmptyMatch"`
+}
+
+type JSONPathAdmissionConfig struct {
+	Rules []JSONPathAdmissionRule `json:"rules", yaml:"rules"`
+}
+
+func init() {
+	admission.RegisterPlugin("jsonpath", func(client clientset.Interface, config io.Reader) (admission.Interface, error) {
+		if config == nil {
+			return nil, errors.New("Config file is required for JSONPath Admission Controller")
+		}
+		jsonConfig := JSONPathAdmissionConfig{}
+		data, err := ioutil.ReadAll(config)
+		if err != nil {
+			return nil, err
+		}
+		jsonData, err := yaml.ToJSON(data)
+		if err != nil {
+			return nil, err
+		}
+		if err := json.Unmarshal(jsonData, &jsonConfig); err != nil {
+			return nil, err
+		}
+		return NewJSONPathAdmission(jsonConfig)
+	})
+}
+
+type jsonPathRule struct {
+	name            string
+	kindRE          *regexp.Regexp
+	path            *util_jsonpath.JSONPath
+	matchRE         *regexp.Regexp
+	acceptNoMatches bool
+}
+
+func (r *jsonPathRule) String() string {
+	return fmt.Sprintf("%s: %s %s %s %v", r.name, r.kindRE.String(), r.path, r.matchRE.String(), r.acceptNoMatches)
+}
+
+func (r *jsonPathRule) admit(a admission.Attributes) error {
+	if !r.kindRE.MatchString(a.GetKind().Kind) {
+		return nil
+	}
+
+	vals, err := r.path.FindResults(a.GetObject())
```

Comment:
xref #27129 and #15390

On Tue, Jun 14, 2016 at 8:49 AM, David Eads <notifications@github.com> wrote:
> In plugin/pkg/admission/jsonpath/admission.go:
>
>> +	kindRE          *regexp.Regexp
>> +	path            *util_jsonpath.JSONPath
>> +	matchRE         *regexp.Regexp
>> +	acceptNoMatches bool
>> +}
>> +
>> +func (r *jsonPathRule) String() string {
>> +	return fmt.Sprintf("%s: %s %s %s %v", r.name, r.kindRE.String(), r.path,
>> r.matchRE.String(), r.acceptNoMatches)
>> +}
>> +
>> +func (r *jsonPathRule) admit(a admission.Attributes) error {
>> +	if !r.kindRE.MatchString(a.GetKind().Kind) {
>> +		return nil
>> +	}
>> +
>> +	vals, err := r.path.FindResults(a.GetObject())
>
> You'll need the external representation of this object to be able to
> evaluate an external json path on it. This is the internal representation.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.